docs/vm-metrics.md

```diff
@@ -0,0 +1,254 @@
+```
+migration_memory_iteration{vm="$UUID"} 0
+vcpu_total_time{vm="$UUID",cpu="0"} 12345
+```
+
+### Implementation
+
+The requisites of the KubeVirt node agent, besides the obvious need to connect to libvirt and
+to expose a prometheus endpoint, are performance and reliability. The collection of metrics
+should require the minimal amount of resources, both as runtime and as container image.
+We know from the experience that shared storage issues may cause VMs to hang, and libvirt
+may in turn hang while collecting metrics from stuck VMs. We want the Node Agent to cope
+gracefully with this scenario, and keep reporting the metrics of the live VMs, skipping
+the stuck VMs. Reporting of stuck VMs is part of health monitoring, which is out of the
+scope of this document.
+
+The options to implement the KubeVirt node agent are:
+
+#### vAdvisor
+We can enhance the [vAdvisor project](https://github.com/KubeVirt/vAdvisor) to implement
+the missing features described above.
+
+**PROS**
+- The project already offers a fair amount of the features we need, it's a great starting point
+- Implemented in python, we can optionally reuse code from the oVirt monitoring
+
+**CONS**
+- Implemented in python, we know from experience in the oVirt project that python is a poor choice
+  scalability-wise. To properly address scale, we will need a rewrite in the future.
+- Implemented in python, requires a relatively big container image
+
+#### vAdvisor, reimplemented
+We can take vAdvisor as starting point and rewrite it in golang - to better integrate with
+the ecosystem right at the start instead of later, and implement the missing features on the
+golang codebase.
+
+**PROS**
+- No scalability issues
+- Better integration with the ecosystem
+- Minimal image size
+
+**CONS**
+- We will pay the rewrite cost upfront. But if we keep the design basics, we can minimize the effort.
+  [Initial experimentations](https://github.com/fromanirh/go-vAdvisor) gave promising results about
+  the development speed
+
+#### collectd
+The [collectd](http://collectd.org/) project fullfills all the KubeVirt monitoring needs, featuring
+a wealth of plugins including a Prometheus endpoint and one libvirt connector.
+Collectd is a well established monitoring solution and it is often mentioned in the examples of
+the Kubernetes documentation about setting a monitoring pipeline.
+
+**PROS**
+- Most of the functionalities ready
+- Established, mature solution
+- Small image size
+
+**CONS**
+- Reporting of ephemeral metrics could need work and acceptance upstream
+- Collectd uses one [implicit and hardcoded mapping](https://github.com/collectd/collectd/blob/master/src/write_prometheus.c#L405)
+  for prometheus labels which is a poor fit for the KubeVirt use case.
+- long release cycle
+
+## KubeVIRT API Adapters: VM metrics naming rules
+
+The task of the KubeVirt HPA API adapter is to expose the metrics collected by the cluster agent
```

Comment:
If HPA stands for Horizontal Pod Autoscaler, I am a little bit confused. HPA is for sure a consumer of Heapster/prometheus metrics, but how is that related to this proposal? Maybe I misunderstand HPA? Or are you thinking about scaling VMs based on that?